### ISSUE TITLE
AIX Specific Issue with Library Paths

### ISSUE THREAD

**Issue by CraigRegester** [intersystems/git-source-control#77]
Seeing a different in how git-source-control (this package) works with IRIS for Health 2021.1 and HealthShare UCR 2021.1 (which uses I4H 2020.1 under the hood.) This is an AIX specific issue but may also impact other *nix systems, not sure?

```
Could not load program /opt/freeware/libexec64/git-core/git-remote-https:
Symbol resolution failed for /opt/freeware/lib64/libcurl.a[libcurl.so.4] because:
	Symbol _GLOBAL__AIXI_libssh2_so (number 4) is not exported from dependent
	  module /ucrfs/sys/healthshare/bin/libssh2.a[libssh2.so.1].
	Symbol _GLOBAL__AIXD_libssh2_so (number 5) is not exported from dependent
	  module /ucrfs/sys/healthshare/bin/libssh2.a[libssh2.so.1].
Examine .loader section symbols with the 'dump -Tv' command
```

[AIX Toolbox](https://www.ibm.com/support/pages/aix-toolbox-linux-applications-downloads-alpha) provides for AIX users to pull down 'safe' versions of tools like git, curl, etc. This works just fine usually but when executing git calls using $ZF from the irisusr/cacheusr-owned process, it prefers its own library path to take precedence. (see module references above)

You can see that git-remote-https doesn't much like that. :) Whereas really, it should be deferring to:

/opt/freeware/lib64/libssh2.a

Because that's the libssh2.a packaged with git/curl that comes from AIX Toolbox. 

I4H 2021.1 I think uses a compatible library so I didn't see this error there (yet) but UCR, based on 2020.1 uses an older libssh2.a and boom, broken. I saw this earlier this year on I4H 2020.1.1 when I was trying to write my own source control hooks using $ZF. My workaround was to inject the proper path. See below. 

```
/* 
		##### Using $zf(-100) #####
		Start magic to get latest git/curl working on AIX 7.2 and using appropriate libssh2 libraries.
		
		Solution is to chain commands together using && (potentially AIX specific) and force library path 
		using the LIBPATH environment variable.
		
		After setting LIBPATH, existing command developed by RC can be executed normally.
	*/
	
	// ZF(-100) supports passing off args as a list. We can take advantage of that to chain commands in the AIX
	// command-line, separating each by the &&. So args(2) and args(4) below are the literals used to split commands
	// and args(3) and args(5) are their own commands that will execute in appropriate order based on their placement.
	// FYI, args(1) is just the first argument for the command executed by $ZF(-100), "echo".
	
	// As we are needing to control library paths being used by git so that HealthShare's out of date libssh2 library
	// isn't utilized and causing the incompatibility, we use LIBPATH (AIX specific environment variable, similar to
	// LD_LIBRARY_PATH in Linux) to force/override the library path the subsequent command in the same shell session
	// will use.
	Set args = 5
	Set args(1) = "Executing Git Command"
	Set args(2) = "&&"
	Set args(3) = "LIBPATH=/usr/lib:/usr/include"
	Set args(4) = "&&"
	Set args(5) = cmd
	
	// Once our args are set, we call out to $ZF(-100) and pass in switches for /SHELL, /NOQUOTE and /STDOUT /STDERR,
	// along with dropping in our args. From there, it's all gravy.
	Set rc=$zf(-100,"/SHELL /NOQUOTE /STDOUT+="""_outputfile_""" /STDERR+="""_outputfile_"""","echo",.args)
```

Wondering if a better approach would be to get $ZF to respect the LibPath variable that can be set under Advanced Memory options (System->Configuration->Advanced Memory Settings) - or providing an option in the Source Control settings. Open to ideas here as you guys would know best!

alternatively, using /SHELL I think forces to pull in the .profile so having the path defined there helps as well.

**Comment by isc-tleavitt** [comment]
I think using /SHELL is probably the simplest solution (and is likely to catch other unrelated gremlins too).

**Comment by isc-tleavitt** [comment]
Also: we must not use /NOQUOTE as this opens the door to command injection.

**Comment by CraigRegester** [comment]
> Also: we must not use /NOQUOTE as this opens the door to command injection.

That's fair - wondered why you guys used the explicit quoting but makes sense now. Don't think I got that far into introspection of its use.

**Comment by isc-tleavitt** [comment]
@CraigRegester my initial reaction to your code was "chaining commands together? how? that shouldn't work..." because I didn't see /NOQUOTE. :)

**Comment by CraigRegester** [comment]
Hmm shell doesn't seem to be taking care of it, not sure it's loading .profile. Tinkering to see what may be happening.

**Comment by CraigRegester** [comment]
Tried setting Path in CPF, LibPath in CPF, shutting down the instance and bringing it back up (per documentation, required for the Path variable) but no matter what, it won't look for libssh2.a in /usr/lib first - /ucrfs/sys/healthshare/bin wins every time.

/SHELL or without, outcome is the same. Any other ISC config setting you guys can think of I can toy with to make it respect the path on a $ZF call? I remember at one point I even explored what it might take to use the 'Callout Gateway'  but that got a little hairy so I dropped it.

OOO for the next few days so won't be able to play with this until next week sometime.

```
DEVUCR:cacheusr:FALIB>Set rc = $zf(-100,"","ldd",.args)
/opt/freeware/libexec64/git-core/git-remote-https needs:
         /usr/lib/libc.a(shr_64.o)
         /usr/lib/libpthreads.a(shr_xpg5_64.o)
         /opt/freeware/lib/libintl.a(libintl.so.8)
         /opt/freeware/lib/libiconv.a(libiconv.so.2)
         /usr/lib/libz.a(libz.so.1)
         /usr/lib/libcurl.a(libcurl.so.4)
         /usr/lib/libexpat.a(libexpat.so.1)
         /unix
         /usr/lib/libcrypt.a(shr_64.o)
         /opt/freeware/lib64/libgcc_s.a(shr.o)
         /ucrfs/sys/healthshare/bin/libssh2.a(libssh2.so.1)
         /opt/freeware/lib/libgssapi_krb5.a(libgssapi_krb5.so)
         /opt/freeware/lib/libkrb5.a(libkrb5.so)
         /opt/freeware/lib/libk5crypto.a(libk5crypto.so)
         /opt/freeware/lib/libcom_err.a(libcom_err.so)
         /opt/freeware/lib/liblber.a(liblber-2.4.so.2)
         /opt/freeware/lib/libldap.a(libldap-2.4.so.2)
         /usr/lib/libssl.a(libssl.so.1.0.2)
         /usr/lib/libcrypto.a(libcrypto.so.1.0.2)
         /opt/freeware/lib64/libkrb5.so
         /opt/freeware/lib64/libk5crypto.so
         /opt/freeware/lib64/libcom_err.so
         /opt/freeware/lib64/libkrb5support.so
         /usr/lib/libintl.a(libintl.so.8)
         /opt/freeware/lib64/liblber.a(liblber-2.4.so.2)
         /usr/lib/libsasl2.a
dump: /usr/lib/libsasl2.a: 0654-108 file is not valid in the current object file mode.
        Use the -X option to specify the desired object mode.
         /usr/lib/libs.a(shr_64.o)
```

**Comment by CraigRegester** [comment]
Last note I'll leave on this one before I head out for the weekend - 

There must have been a fundamental change in $ZF(-100) between IRIS 2020.1 (what UCR uses) and I4H 2021.1 because if I run that ldd command using $ZF(-100) on the same server using the same AIX cacheusr and .profile, I4H 2021.1 respects the pathing natively (using /SHELL or without) even though there is a libssh2.a in /hfs/sys/healthshare/bin (I4H's home)

Good news is I guess this problem may go away once UCR starts utilizing IRIS 2021.1 as its core, bad news is I don't know when that will be.

**Comment by isc-svelury** [comment]
I might be misunderstanding the issue, but would explicitly setting the [core.sshCommand](https://git-scm.com/docs/git-config#Documentation/git-config.txt-coresshCommand) variable with the correct path help?

If so, we could provide a setting for specifying the path of the SSH executable. I am not sure if setting the path this way will work. I haven't been able to get it running on Windows machine owing (I think) to version issues.

**Comment by CraigRegester** [comment]
It's a good thought, I can give it a shot when I get "back to the office".

Think I also heard in one of the virtual summit sessions that ISC is no longer going to be packaging openssl/ssh libraries with their releases as of 2021.2, which would also resolve this sort of issue, but maybe I misinterpreted that.

**Comment by CraigRegester** [comment]
Unfortunately no luck @isc-svelury  - core.sshCommand isn't the crux, it's the dynamically linked libraries beyond that point. 

So  /opt/freeware/libexec64/git-core/git-remote-https is a sub-module of git (packaged with git) that effectively implements curl and it is curl that is then requesting access to libraries like libssh2 and unfortunately, at least when executed from $ZF, IRIS 2020.1.x does not respect any LIBPATH environment variables, whether set in .profile, .kshrc, or in the LibPath CPF variable.

```
DEVUCR:cacheusr:%SYS>set args = 1
DEVUCR:cacheusr:%SYS>set args(1) = "${LIBPATH}"
DEVUCR:cacheusr:%SYS>Set rc = $zf(-100,"/SHELL","echo",.args)
/ucrfs/sys/healthshare/bin   <-- This is where the libssh2.a library lives that we _don't_ want.
```

Little rusty on my AIX Sys Admin stuff but I can see that $ZF is being executed with /usr/bin/ksh (korn shell) and that uses a setuid method that clears LIBPATH with each spawned child process to prevent security holes (someone injecting a bad LIBPATH on the underlying OS)

From what I can tell, IRIS 2021.x (maybe an earlier version as well, not sure) 'fixes' this by not injecting its own LIBPATH on the spawned processes:

```
DEV:cacheusr:%SYS>set args = 1
DEV:cacheusr:%SYS>set args(1) = "${LIBPATH}"
DEV:cacheusr:%SYS>Set rc = $zf(-100,"/SHELL","echo",.args)
     <-- nothing here
```

Therefore OSes' default search path (/etc/environment) takes over and the proper libraries take priority over libraries owned within the InterSystems installation directory.

But unfortunately, this means I'm stuck I think and can't move forward until I can determine a way around this (and I've been fighting it for a few hours today now!) Is there anyway to override what $ZF is doing with LIBPATH using an IRIS environment variable?

**Comment by isc-svelury** [comment]
I couldn't find any references to modifying what $ZF is doing with LIBPATH. @isc-tleavitt do you know of a way this behavior could be changed?

**Comment by isc-tleavitt** [comment]
@CraigRegester if you could engage InterSystems Support on this I think that would be the short route to a solution. Especially since there's apparently a fix between IRIS 2020.1 nd I4H 2021.1 - seems like something an adhoc could fix.

**Comment by isc-tleavitt** [comment]
Aha - @CraigRegester the internal reference for the fix is CDS3257 / DP-13198. It's even in I4H 2020.1.1. I'd say reach out to Support for an adhoc. (Or just wait.)

**Comment by CraigRegester** [comment]
Thanks @isc-tleavitt  - HS 2021.2 was just released and looks like it has I4H 2021.1 baked underneath now (yay!) so may be a moot point now, at least for me :)

**Comment by isc-svelury** [comment]
Closing this issue now.
